packages/medusa/src/services/product.ts

```diff
@@ -0,0 +1,1088 @@
+            include_discount_prices: config.include_discount_prices,
+            region_id: config.region_id,
+          }
+        )
+
+        return [productsWithAdditionalPrices, count]
+      } else {
+        return [products, count]
+      }
+    })
+  }
+
+  /**
+   * Return the total number of documents in database
+   * @param {object} selector - the selector to choose products by
+   * @return {Promise} the result of the count operation
+   */
+  async count(selector: Selector<Product> = {}): Promise<number> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+      /** revisit */
+      const query = buildQuery(selector)
+      return await productRepo.count(query)
+    })
+  }
+
+  /**
+   * Gets a product by id.
+   * Throws in case of DB Error and if product was not found.
+   * @param productId - id of the product to get.
+   * @param config - object that defines what should be included in the
+   *   query response
+   * @return the result of the find one operation.
+   */
+  async retrieve(
+    productId: string,
+    config: FindProductConfig = { include_discount_prices: false }
+  ): Promise<Product> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+      const validatedId = validateId(productId)
+
+      const priceIndex = config.relations?.indexOf("variants.prices") ?? -1
+      if (priceIndex >= 0 && config.relations) {
+        config.relations = [...config.relations]
+        config.relations.splice(priceIndex, 1)
+      }
+
+      const { relations, ...query } = buildQuery({ id: validatedId }, config)
+
+      const product = await productRepo.findOneWithRelations(relations, query)
+
+      if (!product) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `Product with id: ${productId} was not found`
+        )
+      }
+
+      return priceIndex > -1
+        ? await this.setAdditionalPrices(product, {
+            cart_id: config.cart_id,
+            currency_code: config.currency_code,
+            customer_id: config.customer_id,
+            include_discount_prices: config.include_discount_prices,
+            region_id: config.region_id,
+          })
+        : product
+    })
+  }
+
+  /**
+   * Gets a product by handle.
+   * Throws in case of DB Error and if product was not found.
+   * @param productHandle - handle of the product to get.
+   * @param config - details about what to get from the product
+   * @return the result of the find one operation.
+   */
+  async retrieveByHandle(
+    productHandle: string,
+    config: FindProductConfig = {}
+  ): Promise<Product> {
+    return await this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+
+      const priceIndex = config.relations?.indexOf("variants.prices") ?? -1
```

Comment:
See other comment.

suggestion(non-blocking): We are performing this clean-up of relations in quite a lot of places, so it might be worth considering extracting this to a util to slightly improve maintainability.